pkg/backend/openshiftcluster.go

```diff
@@ -197,19 +198,41 @@ func (ocb *openShiftClusterBackend) updateAsyncOperation(ctx context.Context, lo
 			now := time.Now()
 			asyncdoc.AsyncOperation.EndTime = &now
 
+			// Returns a generic error
+			handleOtherError := func() {
+				log.Error(backendErr)
+				asyncdoc.AsyncOperation.Error = &api.CloudErrorBody{
+					Code:    api.CloudErrorCodeInternalServerError,
+					Message: "Internal server error.",
+				}
+			}
+
 			if provisioningState == api.ProvisioningStateFailed {
-				// if type is CloudError - we want to propagate it to the
-				// asyncOperations errors. Otherwise - return generic error
-				err, ok := backendErr.(*api.CloudError)
-				if ok {
+				switch err := backendErr.(type) {
+				case *api.CloudError:
+					// if type is CloudError - we want to propagate it to the asyncOperations errors.
 					log.Print(backendErr)
 					asyncdoc.AsyncOperation.Error = err.CloudErrorBody
-				} else {
-					log.Error(backendErr)
+				case *azure.RequestError:
+					// if it is RequestDisallowedByPolicy, tell user so they can change policy
+					if err.ServiceError != nil && err.ServiceError.Code == api.CloudErrorRequestDisallowedByPolicy {
+						log.Print(backendErr)
+						asyncdoc.AsyncOperation.Error = &api.CloudErrorBody{
+							Code:    api.CloudErrorRequestDisallowedByPolicy,
+							Message: err.ServiceError.Message,
+						}
+					} else {
+						handleOtherError()
+					}
+				case adal.TokenRefreshError:
```

Comment:
What are the circumstances in which this happens and why is it correct to be handling them here?  Note the following:
In GetToken (aad.go:70) we catch token refresh errors and return CloudErrorCodeInvalidServicePrincipalCredentials instead.  The only other obvious place I can see where we might get a token refresh error is in validateActions (openshiftcluster_validatedynamic.go:397), but note that to get to that point we have already successfully called GetToken once via validateServicePrincipalProfile.  So I don't understand what's happening here.

Reply:
Will deal with this in a separate PR